In [2]:
# importing required libraries
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt

In [3]:
# reading the given dataset
train_data = pd.read_csv("training_data.csv")
train_data.head(2)

times     Series1     Series2     Series3     Series4     Series5  \
0  1978-11-01  260.350122  265.379448  285.172883  266.774733  256.704939   
1  1978-11-02  259.943367  273.526405  266.913075  269.664513  265.024105   

      Series6     Series7     Series8     Series9  ...    Series86  \
0  259.962472  263.321555  262.017348  263.854085  ...  258.222624   
1  263.013930  263.091900  254.997818  273.320973  ...  260.759755   

     Series87    Series88    Series89    Series90    Series91    Series92  \
0  274.719836  261.387957  254.903019  268.261197  267.028705  271.567729   
1  256.765620  261.998367  253.702352  259.935957  267.853531  271.902175   

     Series93    Series94    Series95  
0  266.947571  271.930426  271.223367  
1  267.027663  273.179384  261.041906  

[2 rows x 96 columns]

In [4]:
fake_pub_data = pd.read_csv("public_fake_test_data.csv")
fake_pub_data.head(2)

times     Series1     Series2     Series3     Series4     Series5  \
0  1991-02-26  221.906904  209.384404  212.940469  210.951168  200.843305   
1  1991-02-27  221.906904  209.384404  212.940469  210.951168  200.843305   

      Series6     Series7     Series8     Series9  ...    Series86   Series87  \
0  215.455393  203.517218  214.127919  226.265642  ...  213.984178  209.19355   
1  215.455393  203.517218  214.127919  226.265642  ...  213.984178  209.19355   

     Series88   Series89    Series90    Series91   Series92    Series93  \
0  206.859867  216.48676  213.265123  221.629532  217.60735  227.577026   
1  206.859867  216.48676  213.265123  221.629532  217.60735  227.577026   

   Series94    Series95  
0  211.2203  224.895289  
1  211.2203  224.895289  

[2 rows x 96 columns]

In [5]:
def basic_data_understanding(data):
    min_date = data['times'].min()
    max_date = data['times'].max()
    data_info = data.info()
    print("Min Date:", min_date, "|| Max Date:", max_date, "\nData info:\n", data_info)
#     return "Min"min_date, max_date, data_type, check_null

In [6]:
basic_data_understanding(fake_pub_data)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 801 entries, 0 to 800
Data columns (total 96 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   times     801 non-null    object 
 1   Series1   801 non-null    float64
 2   Series2   801 non-null    float64
 3   Series3   801 non-null    float64
 4   Series4   801 non-null    float64
 5   Series5   801 non-null    float64
 6   Series6   801 non-null    float64
 7   Series7   801 non-null    float64
 8   Series8   801 non-null    float64
 9   Series9   801 non-null    float64
 10  Series10  801 non-null    float64
 11  Series11  801 non-null    float64
 12  Series12  801 non-null    float64
 13  Series13  801 non-null    float64
 14  Series14  801 non-null    float64
 15  Series15  801 non-null    float64
 16  Series16  801 non-null    float64
 17  Series17  801 non-null    float64
 18  Series18  801 non-null    float64
 19  Series19  801 non-null    float64
 20  Series20  801 non-null    float6

In [7]:
basic_data_understanding(train_data)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4500 entries, 0 to 4499
Data columns (total 96 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   times     4500 non-null   object 
 1   Series1   4500 non-null   float64
 2   Series2   4500 non-null   float64
 3   Series3   4500 non-null   float64
 4   Series4   4500 non-null   float64
 5   Series5   4500 non-null   float64
 6   Series6   4500 non-null   float64
 7   Series7   4500 non-null   float64
 8   Series8   4500 non-null   float64
 9   Series9   4500 non-null   float64
 10  Series10  4500 non-null   float64
 11  Series11  4500 non-null   float64
 12  Series12  4500 non-null   float64
 13  Series13  4500 non-null   float64
 14  Series14  4500 non-null   float64
 15  Series15  4500 non-null   float64
 16  Series16  4500 non-null   float64
 17  Series17  4500 non-null   float64
 18  Series18  4500 non-null   float64
 19  Series19  4500 non-null   float64
 20  Series20  4500 non-null   floa

### Converting object to timestamp

In [8]:
def convert_timestamp(data):
    if data.dtypes['times'] == object:
        data['times'] = pd.to_datetime(data.times)
        return data 
    else:
        return data

In [9]:
basic_data_understanding(convert_timestamp(train_data))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4500 entries, 0 to 4499
Data columns (total 96 columns):
 #   Column    Non-Null Count  Dtype         
---  ------    --------------  -----         
 0   times     4500 non-null   datetime64[ns]
 1   Series1   4500 non-null   float64       
 2   Series2   4500 non-null   float64       
 3   Series3   4500 non-null   float64       
 4   Series4   4500 non-null   float64       
 5   Series5   4500 non-null   float64       
 6   Series6   4500 non-null   float64       
 7   Series7   4500 non-null   float64       
 8   Series8   4500 non-null   float64       
 9   Series9   4500 non-null   float64       
 10  Series10  4500 non-null   float64       
 11  Series11  4500 non-null   float64       
 12  Series12  4500 non-null   float64       
 13  Series13  4500 non-null   float64       
 14  Series14  4500 non-null   float64       
 15  Series15  4500 non-null   float64       
 16  Series16  4500 non-null   float64       
 17  Series17  4500

### Data basic understanding

**Train Data:** \
-- Min Date: 1978-11-01 00:00:00 || Max Date: 1991-02-25 00:00:00 \
-- Total records: 4500 entries\
-- No missing values in any of the records

**fake_pub_data:** \
-- Min Date: 1991-02-26 || Max Date: 1993-05-06 \
-- Total records: 801 entries\
-- No missing values in any of the records

### Setting time object as index

In [12]:
train_data1 = train_data.set_index('times')
train_data1.head(2)

Series1     Series2     Series3     Series4     Series5  \
times                                                                    
1978-11-01  260.350122  265.379448  285.172883  266.774733  256.704939   
1978-11-02  259.943367  273.526405  266.913075  269.664513  265.024105   

               Series6     Series7     Series8     Series9    Series10  ...  \
times                                                                   ...   
1978-11-01  259.962472  263.321555  262.017348  263.854085  272.871388  ...   
1978-11-02  263.013930  263.091900  254.997818  273.320973  268.203996  ...   

              Series86    Series87    Series88    Series89    Series90  \
times                                                                    
1978-11-01  258.222624  274.719836  261.387957  254.903019  268.261197   
1978-11-02  260.759755  256.765620  261.998367  253.702352  259.935957   

              Series91    Series92    Series93    Series94    Series95  
times                                                                   
1978-11-01  267.028705  271.567729  266.947571  271.930426  271.223367  
1978-11-02  267.853531  271.902175  267.027663  273.179384  261.041906  

[2 rows x 95 columns]

### Weekly aggregation

In [16]:
df_weekly = train_data.groupby(pd.Grouper(key='times',freq='W')).sum()
# df_weekly1 = df_weekly.reset_index(level =['Date'])
df_weekly.head(3)

Series1      Series2      Series3      Series4      Series5  \
times                                                                         
1978-11-05  1304.730170  1338.624324  1350.916401  1317.343844  1180.289057   
1978-11-12  1794.347179  1823.047997  1833.084137  1820.801373  1787.470327   
1978-11-19  1750.186165  1814.972590  1823.850399  1763.190822  1803.962904   

                Series6      Series7      Series8      Series9     Series10  \
times                                                                         
1978-11-05  1300.200680  1306.105590  1297.507665  1184.200970  1378.883843   
1978-11-12  1820.419606  1816.931031  1878.801365  3645.516819  1902.794669   
1978-11-19  1808.079139  1803.050251  1851.496790  1842.466738  1926.178422   

            ...     Series86     Series87     Series88     Series89  \
times       ...                                                       
1978-11-05  ...  1286.825896  1308.711412  1293.667502  1289.824884   
1978-11-12  ...  1830.377237  1757.248391  1684.742270  1831.791919   
1978-11-19  ...  1772.834534  1761.612286  1769.775394  1831.796100   

               Series90     Series91     Series92     Series93     Series94  \
times                                                                         
1978-11-05  1308.813435  1328.431382  1361.895928  1279.894699  1361.870991   
1978-11-12  1829.276312  1869.100983  1889.405363  1833.792927  1864.154420   
1978-11-19  1806.395305  1730.148122  1861.627770  1865.829778  1881.066844   

               Series95  
times                    
1978-11-05  1337.179163  
1978-11-12  1908.670035  
1978-11-19  1887.931503  

[3 rows x 95 columns]

### Monthly aggregation 

In [17]:
df_monthly  = train_data.groupby(pd.Grouper(key='times',freq='M')).sum()
df_monthly.head(3)

Series1      Series2      Series3      Series4      Series5  \
times                                                                         
1978-11-30  7656.715500  7881.887572  7870.854326  7752.660567  7581.901560   
1978-12-31  7875.728987  7874.999700  7872.613723  7899.516216  7663.898783   
1979-01-31  7685.564114  7654.200090  7737.313782  7644.405067  7694.135194   

                Series6      Series7      Series8      Series9     Series10  \
times                                                                         
1978-11-30  7690.017034  7868.409673  7843.973688  9605.083966  8172.032592   
1978-12-31  7740.251913  7780.484120  7708.052874  7922.045601  8149.144537   
1979-01-31  7567.465669  7601.845808  7570.640705  7758.095295  7886.464706   

            ...     Series86     Series87     Series88      Series89  \
times       ...                                                        
1978-11-30  ...  7754.436053  7606.773333  7517.935800   7762.665828   
1978-12-31  ...  7720.968850  7904.003893  7719.780393  10192.002465   
1979-01-31  ...  7494.091124  7576.037977  7587.286562   7753.301233   

               Series90     Series91     Series92     Series93     Series94  \
times                                                                         
1978-11-30  7798.696142  7863.753357  8042.350175  7910.327749  8062.019113   
1978-12-31  7754.513264  8099.610586  7952.848617  8013.071453  8030.335268   
1979-01-31  7478.188333  7842.153085  7644.690077  7950.554917  7804.167895   

               Series95  
times                    
1978-11-30  8082.799024  
1978-12-31  8177.502259  
1979-01-31  7836.404477  

[3 rows x 95 columns]